source/common/router/router.cc

```diff
@@ -739,10 +754,23 @@ void Filter::onResponseTimeout() {
     UpstreamRequestPtr upstream_request =
         upstream_requests_.back()->removeFromList(upstream_requests_);
 
-    // Don't record a timeout for upstream requests we've already seen headers
-    // for.
+    // Don't do work for upstream requests we've already seen headers for.
     if (upstream_request->awaiting_headers_) {
       cluster_->stats().upstream_rq_timeout_.inc();
+
+      if (cluster_->timeoutBudgetStats().has_value()) {
+        // Cancel firing per-try timeout information, because the per-try timeout did not come into
+        // play when the global timeout was hit.
+        upstream_request->timeout_budget_stats_->cancel();
+
+        Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+        std::chrono::milliseconds response_time =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
```

Comment:
In this case, you know the timeout as it was the global timeout, right? Can you just use it directly to be more clear?

Reply:
Yes, this should just fire 100.